aten/src/TH/generic/THTensorApply.hpp

```diff
@@ -1,41 +1,323 @@
+
 #ifndef NAN
   #define NAN (nan(NULL))
 #endif
 
-#ifdef _OPENMP
-#include <omp.h>
-#endif
-
 #define HYPER_TH_OMP_OVERHEAD_THRESHOLD 2000
 #define ORDIN_TH_OMP_OVERHEAD_THRESHOLD 20000
 #define UNCERTAIN_TH_OMP_OVERHEAD_THRESHOLD 50000
 #define TH_OMP_OVERHEAD_THRESHOLD 100000
 
-#ifdef _OPENMP
+#define TH_CHECK_SAME_SIZE(TENSOR1, TENSOR2) \
+{ \
+  if (!THTensor_(isSameSizeAs)(TENSOR1, TENSOR2)) { \
+    AT_ERROR("inconsistent tensor size, expected ", #TENSOR1, " ", TENSOR1->sizes(), " and ", #TENSOR2, " ", TENSOR2->sizes(), " to have the same size"); \
+  } \
+}
+
+// Used for `scatter` and `scatterAdd`
+// Assumes TENSOR1 is real
+//         TENSOR2 is src
+//         TENSOR3 is index
+// Tests:
+//   1. index->size(d) <= src->size(d) for all d
+//   2. index->size(d) <= real->size(d) for all d != dim
+#define TH_TENSOR_DIM_APPLY3_SIZE_SCATTER(TENSOR1, TENSOR2, TENSOR3, DIMENSION) \
+{ \
+  int shape_check_flag = 0; \
+  for (TH_TENSOR_DIM_APPLY_i = 0; TH_TENSOR_DIM_APPLY_i < THTensor_nDimensionLegacyAll(TENSOR1); TH_TENSOR_DIM_APPLY_i++) \
+  { \
+    int64_t TENSOR3##_dim_size = THTensor_sizeLegacyNoScalars(TENSOR3, TH_TENSOR_DIM_APPLY_i); \
+    if (TH_TENSOR_DIM_APPLY_i != DIMENSION) { \
+      if (TENSOR3##_dim_size > THTensor_sizeLegacyNoScalars(TENSOR1, TH_TENSOR_DIM_APPLY_i)) { \
+        shape_check_flag = 1; \
+        break; \
+      } \
+    } \
+    if (TENSOR3##_dim_size > THTensor_sizeLegacyNoScalars(TENSOR2, TH_TENSOR_DIM_APPLY_i)) { \
+      shape_check_flag = 1; \
+      break; \
+    } \
+  } \
+  if (shape_check_flag == 1) { \
+    AT_ERROR("Expected ", #TENSOR3, " ", TENSOR3->sizes(), " to be smaller size than ", #TENSOR2, " ", TENSOR2->sizes(), " and to be smaller than ", #TENSOR1, " ", TENSOR1->sizes(), " apart from dimension ", DIMENSION); \
+  } \
+}
+
+#undef th_isnan
+#if defined(TH_REAL_IS_FLOAT) || defined(TH_REAL_IS_DOUBLE)
+#define th_isnan(val) \
+(std::isnan(val))
+#else
+#define th_isnan(val) (0)
+#endif
+
+#undef th_isnan_break
+#if defined(TH_REAL_IS_FLOAT) || defined(TH_REAL_IS_DOUBLE)
+#define th_isnan_break(val) \
+if (std::isnan(val)) break;
+#else
+#define th_isnan_break(val)
+#endif
 
 #ifdef _WIN32
-// MSVC doesing support loop pragmas, but does support others. Create a new macro to account for those differences.
+// MSVC doesn't support loop pragmas, but does support others. Create a new macro to account for those differences.
 #define PRAGMA_LOOP(P)    // Noop
-#define PRAGMA(P)         __pragma(P)
 #else
 #define PRAGMA_LOOP(P)    _Pragma(#P)
-#define PRAGMA(P)         _Pragma(#P)
 #endif
 
+#define TH_TENSOR_APPLY2_PARALLEL(SIZE, CONTIG1, CONTIG2, TYPE1, TENSOR1, TYPE2, TENSOR2, CODE, THRESHOLD) \
+{ \
+  /* for advanced searching index*/ \
+  if (CONTIG1 && CONTIG2) { \
+    TYPE1 *rp = THTensor_getStoragePtr(TENSOR1)->data<TYPE1>()+TENSOR1->storage_offset(); \
+    TYPE2 *tp = THTensor_getStoragePtr(TENSOR2)->data<TYPE2>()+TENSOR2->storage_offset(); \
+    if (tp != (TYPE2*)rp) { \
```

Comment:
I wonder if we can go a step further for these and start using the TensorIterator. We should probably do that in a follow-up PR however, unless it can significantly simplify this code.

Reply:
cc @VitalyFedyunin

Reply:
One of the risks however is that we're starting to introduce too many dependencies on ATen. One of the opportunities is to move a lot of these ops into an e.g. legacy_implementation.cpp within aten/native and delete them from TH all-together. They can then be refactored into something nicer down the road.

Reply:
tbh my goal in this PR was purely porting to ATen/Parallel primitive without much changes to TH, to speed up transition away from OpenMP and minimize chance of introducing new bugs. We can think about using TensorIterator in the next PRs, or deprecate TH altogether